website_sale_preset_carts/controllers/suspend_cart.py

```diff
@@ -0,0 +1,127 @@
+                )
+                return request.redirect(
+                    request.params.get("redirect")
+                )
+        qcontext = {
+            "data": data,
+            "initial": self.suspend_cart_form_initial(context=form_context),
+            "errors": errors,
+            "redirect": request.params.get("redirect") or "/my/home",
+            "today": date.today().isoformat(),
+        }
+        return request.render(
+            "website_sale_preset_carts.suspend_cart_form", qcontext
+        )
+
+    def suspend_cart_form_initial(self, context=None):
+        """Return initial for suspend cart form."""
+        context = {} if context is None else context
+        initial = {}
+        partner = context.get("partner")
+        if partner.suspend_cart:
+            initial["date_from"] = partner.cart_suspended_from
+            initial["date_to"] = partner.cart_suspended_date
+        else:
+            initial["date_from"] = date.today().isoformat()
+            initial["date_to"] = (date.today() + timedelta(days=1)).isoformat()
+        return initial
+
+    def clean_and_validate_form(self, data):
+        """Return the cleaned data and errors."""
+        cleaned_data = {}
+        errors = {}
+        # Date from
+        field = "date_from"
+        value = data.get(field)
+        try:
+            value = self.clean_and_validate_date_from(value, data=data)
+            cleaned_data[field] = value
+        except ValueError as err:
+            errors.update({field: err})
+        # Date to
+        field = "date_to"
+        value = data.get(field)
+        try:
+            value = self.clean_and_validate_date_from(value, data=data)
+            cleaned_data[field] = value
+        except ValueError as err:
+            errors += {field: err}
+        return cleaned_data, errors
+
+    def clean_and_validate_date_from(self, value, data=None):
+        """
+        Return the cleaned value and Raise ValueError if the
+        date_from is not valid.
+        """
+        if not value:
+            raise ValueError(_("This field is required."))
+        try:
+            value_date = datetime.strptime(value, "%Y-%m-%d").date()
+        except ValueError:
+            raise ValueError(_("Wrong format for the date."))
+        if value_date <= date.today():
+            raise ValueError(_("Date can not be in the past."))
+        return value
+
+    def clean_and_validate_date_to(self, value, data=None):
+        """
+        Return the cleaned value and Raise ValueError if the
+        date_to is not valid.
+        """
+        if not value:
+            raise ValueError(_("This field is required."))
+        try:
+            value_date = datetime.strptime(value, "%Y-%m-%d").date()
+        except ValueError:
+            raise ValueError(_("Wrong format for the date."))
+        try:
+            date_from = datetime.strptime(
+                data.get("date_from"), "%Y-%m-%d"
```

Comment:
Although in this module the param `data=data` is systematically given, I think a `AttributeError: 'NoneType'` can be raised if this method is used without it

Reply:
Right. As I use data param, it should be mandatory and not optional. Or as you suggest a test should be done to prevent a data.get() on a None Value.